File: service/control_flow_graph.py
import ast


class _Block:
    def __init__(self, block_id, hint, statements=None):
        self.statements = statements or []
        self.block_id = block_id
        self.hint = hint

    def __repr__(self):
        import codegen
        code = '\n'.join(['    ' + codegen.to_source(stmt) for stmt in self.statements])
        return "BLOCK " + str(self.block_id) + " (" + self.hint + "):\n" + code


class ControlFlowGraph:

    def __init__(self, statements):
        self.global_id = 0
        self.blocks = []
        self.successors = set()
        exception_exit_block = self._make_block('exceptional exit')
        self.entry, self.exit = self._cfg_for_statements(
            'entry', statements, None, None, exception_exit_block)

    def _make_block(self, hint: str, statements=None) -> _Block:
        block = _Block(self.global_id, hint, statements)
        self.blocks.append(block)
        self.global_id += 1
        return block

    def get_blocks(self):
        """
        Get the basic blocks in the control flow graph, starting with
        the entry block.
        """
        visited = set()
        to_visit = set([self.entry])
        while to_visit:
            block = to_visit.pop()
            visited.add(block)
            yield block
            for (pred, succ) in self.successors:
                if pred == block and not (succ in visited):
                    to_visit.add(succ)

    def get_successors(self, block):
        """
        Get the blocks that can be executed directly after the given basic block.
        """
        return [succ for (pred, succ) in self.successors if pred == block]

    def get_predecessors(self, block):
        """
        Get the blocks that can be executed directly before the given basic block.
        """
        return [pred for (pred, succ) in self.successors if succ == block]

    def print(self):
        """
        Print the control flow graph, for debugging.
        """
        print('CFG', 'ENTRY:', self.entry.block_id, 'EXIT:', self.exit.block_id)
        for block in self.get_blocks():
            print(block.__repr__())
            if block == self.exit:
                print('    EXIT')
            else:
                print('    SUCC', [succ.block_id for (pred, succ)
                                   in self.successors if pred == block])

    def _add_successor(self, block1, block2):
        self.successors.add((block1, block2))

    def _handle_if(self, statement: ast.If, last, closest_loop_head, closest_loop_exit, exc_block):
        test_block = self._make_block('if test', [statement.test])
        self._add_successor(last, test_block)
        body_entry, body_exit = self._cfg_for_statements(
            'if body', statement.body, closest_loop_head, closest_loop_exit, exc_block)
        self._add_successor(test_block, body_entry)
        join_block = self._make_block('conditional join')
        self._add_successor(body_exit, join_block)
        if statement.orelse:
            else_entry, else_exit = self._cfg_for_statements(
                'else body', statement.orelse, closest_loop_head, closest_loop_exit, exc_block)
            self._add_successor(test_block, else_entry)
            self._add_successor(else_exit, join_block)
        else:
            self._add_successor(test_block, join_block)
        return join_block

    def _handle_while(self, statement: ast.While, last, exc_block):
        head = [statement.test]
        loop_head_block = self._make_block('while loop head', head)
        self._add_successor(last, loop_head_block)
        after_loop = self._make_block('while loop join')
        body_entry, body_exit = self._cfg_for_statements(
            'while body', statement.body, loop_head_block, after_loop, exc_block)
        self._add_successor(loop_head_block, body_entry)
        self._add_successor(body_exit, loop_head_block)  # back edge
        self._add_successor(loop_head_block, after_loop)
        return after_loop

    def _handle_for(self, statement: ast.For, last, exc_block):
        # We approximate an iter call with an assignment
        head = [ast.Assign(statement.target, statement.iter)]
        loop_head_block = self._make_block('for loop head', head)
        self._add_successor(last, loop_head_block)
        after_loop = self._make_block('for loop join')
        body_entry, body_exit = self._cfg_for_statements(
            'for body', statement.body, loop_head_block, after_loop, exc_block)
        self._add_successor(loop_head_block, body_entry)
        self._add_successor(body_exit, loop_head_block)  # back edge
        self._add_successor(loop_head_block, after_loop)
        return after_loop

    def _handle_with(self, statement: ast.With, last, closest_loop_head, closest_loop_exit, exc_block):
        resource_block = self._make_block([statement.items])
        self._add_successor(last, resource_block)
        body_entry, body_exit = self._cfg_for_statements(
            'with body', statement.body, closest_loop_head, closest_loop_exit, exc_block)
        self._add_successor(resource_block, body_entry)
        return body_exit

    def _handle_try(self, statement: ast.Try, last, closest_loop_head, closest_loop_exit, exc_block):
        # a block for normal exit from the body
        after_try = self._make_block('try join')
        # For the handlers, create a block that fans out to all handler bodies.
        # This block is reachable from any nested raise statement.
        # FIXME make this block reachable from function calls
        if statement.handlers:
            handler_head = self._make_block('handler fan out')
            handler_pairs = [self._cfg_for_statements(
                'handler body', h.body, closest_loop_head, closest_loop_exit, exc_block) for h in statement.handlers]
            for (entry, _) in handler_pairs:
                self._add_successor(handler_head, entry)
        # Translate the body, using the new exception block.
        body_entry, body_exit = self._cfg_for_statements(
            'try body', statement.body, closest_loop_head, closest_loop_exit,
            handler_head if statement.handlers else exc_block)
        self._add_successor(last, body_entry)
        # The finally block comes after the body and any handlers,
        # with the original exception context.
        if statement.finalbody:
            finally_entry, finally_exit = self._cfg_for_statements(
                'finally body', statement.finalbody,  closest_loop_head, closest_loop_exit, exc_block)
            self._add_successor(body_exit, finally_entry)
            self._add_successor(finally_exit, after_try)
            # All the handler CFGs go to the finally.
            if statement.handlers:
                for (_, handler_exit) in handler_pairs:
                    self._add_successor(handler_exit, finally_entry)
        else:
            # All the handler CFGs go to the statement after the try,
            # since the corresponding exception was handled.
            if statement.handlers:
                for (_, handler_exit) in handler_pairs:
                    self._add_successor(handler_exit, after_try)
            self._add_successor(body_exit, after_try)
        return after_try

    def _cfg_for_statements(self, hint, statements, closest_loop_head, closest_loop_exit, exc_block):
        entry = self._make_block(hint)
        last = entry
        for statement in statements:
            if isinstance(statement, ast.If):
                last = self._handle_if(
                    statement, last, closest_loop_head, closest_loop_exit, exc_block)
            elif isinstance(statement, ast.While):
                last = self._handle_while(statement, last, exc_block)
            elif isinstance(statement, ast.For):
                last = self._handle_for(statement, last, exc_block)
            elif isinstance(statement, ast.With):
                last = self._handle_with(
                    statement, last, closest_loop_head, closest_loop_exit, exc_block)
            elif isinstance(statement, ast.Try):
                last = self._handle_try(
                    statement, last, closest_loop_head, closest_loop_exit, exc_block)
            elif isinstance(statement, ast.Raise):
                self._add_successor(last, exc_block)
            elif isinstance(statement, ast.Break):
                self._add_successor(last, closest_loop_exit)
            elif isinstance(statement, ast.Continue):
                self._add_successor(last, closest_loop_head)
            else:
                last.statements.append(statement)
        return (entry, last)

File: src/ControlFlowGraph.ts
import * as ast from './parsers/python/python_parser';
import { Set } from './Set';


export class Block {

    constructor(
        public id: number,
        private hint: string,
        public statements: ast.ISyntaxNode[],
        public loopVariables: ast.ISyntaxNode[] = []) {
    }

    public toString(): string {
        return 'BLOCK ' + this.id + ' (' + this.hint + ')\n' +
            this.statements.map(s => '    line ' + s.location.first_line).join('\n');
    }
}


class BlockSet extends Set<Block> {
    constructor(...items: Block[]) {
        super(b => b.id.toString(), ...items);
    }
}


class Context {
    constructor(
        public loopHead: Block,
        public loopExit: Block,
        public exceptionBlock: Block) {
    }
    public forLoop(loopHead: Block, loopExit: Block): Context {
        return new Context(loopHead, loopExit, this.exceptionBlock);
    }
    public forExcepts(exceptionBlock: Block): Context {
        return new Context(this.loopHead, this.loopExit, exceptionBlock);
    }
}


export class ControlFlowGraph {
    private _blocks: Block[] = [];
    private globalId = 0;
    private entry: Block;
    private exit: Block;
    private successors = new Set<[Block, Block]>(([b1, b2]) => b1.id + ',' + b2.id);
    private loopVariables: ast.ISyntaxNode[][] = [];

    constructor(module: ast.IModule) {
        [this.entry, this.exit] = this.makeCFG(
            'entry', module.code, new Context(null, null, this.makeBlock('exceptional exit')));
    }

    private makeBlock(hint: string, statements: ast.ISyntaxNode[] = []) {
        const b = new Block(this.globalId++, hint, statements);
        if (this.loopVariables.length) {
            b.loopVariables = this.loopVariables[this.loopVariables.length - 1];
        }
        this._blocks.push(b);
        return b;
    }

    public get blocks(): Block[] {
        const visited: Block[] = [];
        const toVisit = new BlockSet(this.entry);
        while (!toVisit.empty) {
            const block = toVisit.take();
            visited.push(block);
            this.successors.items.forEach(([pred, succ]) => {
                if (pred === block && visited.indexOf(succ) < 0) {
                    toVisit.add(succ);
                }
            });
        }
        return visited;
    }

    public getSuccessors(block: Block): Block[] {
        return this.successors.items
            .filter(([p, _]) => p == block)
            .map(([_, s]) => s);
    }

    public getPredecessors(block: Block): Block[] {
        return this.successors.items
            .filter(([_, s]) => s == block)
            .map(([p, _]) => p);
    }

    public print() {
        console.log('CFG', 'ENTRY:', this.entry.id, 'EXIT:', this.exit.id);
        this.blocks.forEach(block => {
            console.log(block.toString());
            if (block === this.exit) {
                console.log('    EXIT');
            } else {
                console.log('    SUCC',
                    this.getSuccessors(block).map(b => b.id.toString()).join(','));
            }
        });
    }

    private link(...blocks: Block[]): void {
        for (let i = 1; i < blocks.length; i++)
            this.successors.add([blocks[i - 1], blocks[i]]);
    }

    private handleIf(statement: ast.IIf, last: Block, context: Context): Block {
        const ifCondBlock = this.makeBlock('if cond', [statement.cond]);
        const [bodyEntry, bodyExit] = this.makeCFG('if body', statement.code, context);
        this.link(last, ifCondBlock, bodyEntry);
        const joinBlock = this.makeBlock('conditional join');
        this.link(bodyExit, joinBlock);
        let lastCondBlock: Block = ifCondBlock;
        if (statement.elif) {
            statement.elif.forEach(elif => {
                const elifCondBlock = this.makeBlock('elif cond', [elif.cond]);
                this.link(lastCondBlock, elifCondBlock);
                const [elifEntry, elifExit] = this.makeCFG('elif body', elif.code, context);
                this.link(elifCondBlock, elifEntry);
                this.link(elifExit, joinBlock);
                lastCondBlock = elifCondBlock;
            });
        }
        if (statement.else && statement.else.length) {
            const [elseEntry, elseExit] = this.makeCFG('else body', statement.else, context);
            this.link(lastCondBlock, elseEntry);
            this.link(elseExit, joinBlock);
        } else {
            this.link(lastCondBlock, joinBlock);
        }
        return joinBlock;
    }

    private handleWhile(statement: ast.IWhile, last: Block, context: Context): Block {
        const loopHeadBlock = this.makeBlock('while loop head', [statement.cond]);
        this.link(last, loopHeadBlock);
        const afterLoop = this.makeBlock('while loop join');
        this.loopVariables.push([statement.cond]);
        const [bodyEntry, bodyExit] = this.makeCFG('while body', statement.code, context.forLoop(loopHeadBlock, afterLoop));
        this.loopVariables.pop();
        this.link(loopHeadBlock, bodyEntry);
        this.link(bodyExit, loopHeadBlock); // back edge
        this.link(loopHeadBlock, afterLoop);
        return afterLoop;
    }

    private handleFor(statement: ast.IFor, last: Block, context: Context): Block {
        const loopHeadBlock = this.makeBlock('for loop head',
            // synthesize a statement to simulate using the iterator
            [{ type: ast.ASSIGN, op: undefined, sources: statement.iter, targets: statement.target, location: statement.iter[0].location }]);
        this.link(last, loopHeadBlock);
        const afterLoop = this.makeBlock('for loop join');
        this.loopVariables.push(statement.target);
        const [bodyEntry, bodyExit] = this.makeCFG('while body', statement.code, context.forLoop(loopHeadBlock, afterLoop));
        this.loopVariables.pop();
        this.link(loopHeadBlock, bodyEntry);
        this.link(bodyExit, loopHeadBlock); // back edge
        this.link(loopHeadBlock, afterLoop);
        return afterLoop;
    }

    private handleWith(statement: ast.IWith, last: Block, context: Context): Block {
        const assignments = statement.items.map(
            ({ with: w, as: a }) => (<ast.IAssignment>{ type: ast.ASSIGN, targets: [a], sources: [w], location: w.location }))
        const resourceBlock = this.makeBlock('with', assignments);
        this.link(last, resourceBlock);
        const [bodyEntry, bodyExit] = this.makeCFG('with body', statement.code, context);
        this.link(resourceBlock, bodyEntry);
        return bodyExit;
    }

    private handleTry(statement: ast.ITry, last: Block, context: Context): Block {
        const afterTry = this.makeBlock('try join');
        let exnContext = context;
        let handlerExits: Block[] = [];
        if (statement.excepts) {
            const handlerHead = this.makeBlock('handlers');
            const handlerCfgs = statement.excepts.map(
                handler => this.makeCFG('handler body', handler.code, context));
            handlerCfgs.forEach(([exceptEntry, _]) => this.link(handlerHead, exceptEntry));
            exnContext = context.forExcepts(handlerHead);
            handlerExits = handlerCfgs.map(([_, exceptExit]) => exceptExit);
        }
        const [bodyEntry, bodyExit] = this.makeCFG('try body', statement.code, exnContext);
        this.link(last, bodyEntry);
        let normalExit = bodyExit;
        if (statement.else) {
            const [elseEntry, elseExit] = this.makeCFG('try else body', statement.else, context);
            this.link(normalExit, elseEntry);
            normalExit = elseExit;
        }
        if (statement.finally) {
            const [finallyEntry, finallyExit] = this.makeCFG('finally body', statement.finally, context);
            this.link(normalExit, finallyEntry);
            this.link(finallyExit, afterTry);
            handlerExits.forEach(handlerExit => this.link(handlerExit, finallyEntry));
        } else {
            handlerExits.forEach(handlerExit => this.link(handlerExit, afterTry));
            this.link(normalExit, afterTry);
        }
        return afterTry;
    }

    private makeCFG(hint: string, statements: ast.ISyntaxNode[], context: Context): [Block, Block] {
        const entry = this.makeBlock(hint);
        let last = entry;
        statements.forEach(statement => {
            switch (statement.type) {
                case ast.DEF:
                    break;
                case ast.IF:
                    last = this.handleIf(statement, last, context);
                    break;
                case ast.WHILE:
                    last = this.handleWhile(statement, last, context);
                    break;
                case ast.FOR:
                    last = this.handleFor(statement, last, context);
                    break;
                case ast.WITH:
                    last = this.handleWith(statement, last, context);
                    break;
                case ast.TRY:
                    last = this.handleTry(statement, last, context);
                    break;
                case ast.RAISE:
                    this.link(last, context.exceptionBlock);
                    return;
                case ast.BREAK:
                    this.link(last, context.loopExit);
                    return;
                case ast.CONTINUE:
                    this.link(last, context.loopHead);
                    return;
                default:
                    last.statements.push(statement);
                    break;
            }
        });
        return [entry, last];
    }
}
